File: data/tree-height-of-a-binary-tree.cpp
// Tree: Height of a Binary Tree
// Given a binary tree, print its height.
//
// https://www.hackerrank.com/challenges/tree-height-of-a-binary-tree/problem
//

#include <iostream>
#include <cstddef>

using namespace std;

class Node{
    public:
        int data;
        Node *left;
        Node *right;
        Node(int d){
            data = d;
            left = NULL;
            right = NULL;
        }
};

class Solution {
    public:
  		Node* insert(Node* root, int data) {
            if(root == NULL) {
                return new Node(data);
            }
            else {
                Node* cur;
                if(data <= root->data){
                    cur = insert(root->left, data);
                    root->left = cur;
                }
                else{
                    cur = insert(root->right, data);
                    root->right = cur;
               }

               return root;
           }
        }
// (skeliton_head) ----------------------------------------------------------------------

/*The tree node has data, left child and right child
class Node {
    int data;
    Node* left;
    Node* right;
};

*/

    int height(Node* root, int level = 0)
    {
        int h = level;
        if (root->left) h = std::max(h, height(root->left, level + 1));
        if (root->right) h = std::max(h, height(root->right, level + 1));
        return h;
    }

// (skeliton_tail) ----------------------------------------------------------------------
}; //End of Solution

int main() {
    Solution myTree;
    Node* root = NULL;
    int t;
    int data;

    cin >> t;

    while(t-- > 0){
        cin >> data;
        root = myTree.insert(root, data);
    }
    int height = myTree.height(root);
    cout << height;

    return 0;
}

File: data-structures/trees/tree-height-of-a-binary-tree.py
# Tree: Height of a Binary Tree
# Given a binary tree, print its height.
#
# https://www.hackerrank.com/challenges/tree-height-of-a-binary-tree/problem
#

class Node:
    def __init__(self,info):
        self.info = info
        self.left = None
        self.right = None
        self.level = None

    def __str__(self):
        return str(self.info)

class BinarySearchTree:
    def __init__(self):
        self.root = None

    def create(self, val):
        if self.root == None:
            self.root = Node(val)
        else:
            current = self.root

            while True:
                if val < current.info:
                    if current.left:
                        current = current.left
                    else:
                        current.left = Node(val)
                        break
                elif val > current.info:
                    if current.right:
                        current = current.right
                    else:
                        current.right = Node(val)
                        break
                else:
                    break
# (skeliton_head) ----------------------------------------------------------------------

# Enter your code here. Read input from STDIN. Print output to STDOUT
'''
class Node:
      def __init__(self,info):
          self.info = info
          self.left = None
          self.right = None


       // this is a node of the tree , which contains info as data, left , right
'''

    def getHeight(self,root,height=0):
        #Write your code here

        h = height
        if root.left is not None:
            h = max(h, self.getHeight(root.left, height + 1))
        if root.right is not None:
            h = max(h, self.getHeight(root.right, height + 1))
        return h


# (skeliton_tail) ----------------------------------------------------------------------
tree = BinarySearchTree()
t = int(input())

for _ in range(t):
    x = int(input())
    tree.create(x)

print(height(tree.root))
